src/lib/jobrunner.rs

```diff
@@ -0,0 +1,409 @@
+        // there's no likelihood that we can recover.
+
+        loop {
+            let pjob = self.snare.queue.lock().unwrap().pop();
+            match pjob {
+                Some(qj) => {
+                    if self.num_running == self.snare.config.maxjobs {
+                        self.snare.queue.lock().unwrap().push_front(qj);
+                        return false;
+                    }
+                    match self.try_job(qj) {
+                        Ok(j) => {
+                            // The unwrap is safe since we've already checked that there's room to
+                            // run at least 1 job.
+                            let i = self.running.iter().position(|x| x.is_none()).unwrap();
+                            self.running[i] = Some(j);
+                            self.num_running += 1;
+                            self.update_pollfds();
+                        }
+                        Err(Some(qj)) => {
+                            // The job couldn't be run for temporary reasons: we'll retry later.
+                            self.snare.queue.lock().unwrap().push_front(qj);
+                            return false;
+                        }
+                        Err(None) => {
+                            // The job couldn't be run for permanent reasons: it has been consumed
+                            // and can't be rerun. Perhaps surprisingly, this is equivalent to the
+                            // job having run successfully: since it hasn't been put back on the
+                            // queue, there's no need to tell the caller that we couldn't pop all
+                            // the jobs on the queue.
+                        }
+                    }
+                }
+                None => return true,
+            }
+        }
+    }
+
+    /// Try starting the `QueueJob` `qj` running, returning `Ok(Job)` upon success. If for
+    /// temporary reasons that is not possible, the job is returned via `Err(Some(QueueJob))` so
+    /// that it can be put back in the queue and retried later. If `Err(None)` is returned then the
+    /// job could not be run for permanent reasons and the job is consumed.
+    fn try_job(&mut self, qj: QueueJob) -> Result<Job, Option<QueueJob>> {
+        // Write the JSON to an unnamed temporary file.
+        let json_path = match NamedTempFile::new() {
+            Ok(tfile) => match tfile.into_temp_path().keep() {
+                Ok(p) => {
+                    if fs::write(&p, qj.json_str.as_bytes()).is_err() {
+                        remove_file(p).ok();
+                        return Err(Some(qj));
+                    }
+                    p
+                }
+                Err(_) => return Err(Some(qj)),
+            },
+            Err(_) => return Err(Some(qj)),
+        };
+
+        // We combine the child processe's stderr/stdout and write them to an unnamed temporary
```

Comment:
Is that a typo `processe's`?  Not sure myself.

Plural of process is "processes", so I'd guess *processes'*?

Reply:
Fixed in https://github.com/softdevteam/snare/pull/7/commits/6283b7a29a721248f11d0b430033df2bb7cb4eec.

Reply:
So now we have another variation: `process's`. I don't know which is right, but to be clear that's another variation to what was suggested above.

It's my understanding that to make a word that already ends with an `s` posessive, you add a `'`.

But as you know, I'm not good at grammar!

Reply:
I refer you to http://www.apostrophe.org.uk/page2.html -- which is what I always have to check as I can never remember either!

Reply:
Ah! Cool!

But wait!

> ... however, if there are two or more dogs, companies or Joneses in our example, the apostrophe comes after the 's':
>
>    the dogs' bones
>    the companies' logos
>    Joneses' bakeries

 * Singular *process*
 * Plural *processes*
 * Possessive plural *processes'*

(I know, this is getting absurd :P )

Reply:
Wait, I see what's going on here.

We are combining stdout and stderr for one process, not the stdout and stderr for many processes. So you are right!